File: sdk/lx-trading-cpp/examples/arbitrage_bot.cpp
/**
 * LX-First Arbitrage Bot Example
 *
 * This bot uses the LX-First strategy where LX DEX prices are treated
 * as the "truth" (fastest venue with nanosecond updates, 200ms blocks).
 * Other venues are always stale by comparison.
 *
 * Arbitrage = exploiting stale venues before they catch up to LX prices.
 *
 * Cross-chain transport options:
 * - Warp: For Lux subnet communication (<500ms)
 * - Teleport: For EVM chain bridging (~30s)
 * - CEX API: Direct trading (instant)
 *
 * NO SMART CONTRACTS - just coordinated trades through unified SDK.
 */

#include <lx/trading/lx_dex.hpp>
#include <lx/trading/arbitrage/arbitrage.hpp>
#include <iostream>
#include <thread>
#include <atomic>
#include <chrono>
#include <random>
#include <csignal>
#include <mutex>

using namespace lx::trading;
using namespace lx::trading::arbitrage;

// ============================================
// Global state
// ============================================

std::atomic<bool> g_running{false};
std::mutex g_stats_mutex;

struct Stats {
    uint64_t total_opportunities = 0;
    uint64_t total_executions = 0;
    Decimal total_pnl{"0"};
};

Stats g_stats;

// ============================================
// Arbitrage Bot Class
// ============================================

class ArbitrageBot {
public:
    ArbitrageBot() = default;

    bool start() {
        std::cout << std::string(60, '=') << std::endl;
        std::cout << "LX-FIRST ARBITRAGE BOT" << std::endl;
        std::cout << std::string(60, '=') << std::endl;
        std::cout << std::endl;

        // Initialize DEX client
        LxDexConfig dex_config;
        dex_config.endpoint = get_env("LX_DEX_ENDPOINT", "wss://dex.lux.network/ws");
        dex_config.api_key = get_env("LX_API_KEY", "");

        dex_ = std::make_shared<LxDex>(dex_config);
        if (!dex_->connect()) {
            std::cerr << "Failed to connect to LX DEX" << std::endl;
            return false;
        }
        std::cout << "[OK] Connected to LX DEX" << std::endl;

        // Initialize LX-First strategy
        LxFirstConfig lx_config;
        lx_config.max_staleness_ms = 2000;
        lx_config.min_divergence_bps = Decimal("10");
        lx_config.min_profit = Decimal("5");
        lx_config.max_position_size = Decimal("10000");
        lx_config.symbols = {"BTC-USDC", "ETH-USDC", "LUX-USDC"};
        lx_config.venue_latencies = {
            {"binance", 50},
            {"mexc", 100},
            {"okx", 80},
            {"uniswap", 12000},
            {"pancakeswap", 3000}
        };

        lx_first_ = std::make_shared<LxFirstArbitrage>(lx_config);
        lx_first_->on_opportunity([this](const LxFirstOpportunity& opp) {
            on_lx_first_opportunity(opp);
        });
        std::cout << "[OK] LX-First strategy initialized" << std::endl;

        // Initialize Scanner
        ScannerConfig scanner_config;
        scanner_config.min_spread_bps = Decimal("10");
        scanner_config.min_profit_usd = Decimal("10");
        scanner_config.max_price_age_ms = 5000;
        scanner_config.symbols = {"BTC", "ETH", "LUX", "SOL", "AVAX"};
        scanner_config.chain_ids = {"lux", "ethereum", "bsc", "arbitrum", "polygon"};
        scanner_config.scan_interval_ms = 100;

        scanner_ = std::make_shared<Scanner>(scanner_config);
        scanner_->on_opportunity([](const ArbitrageOpportunity& opp) {
            std::cout << "[SCANNER] " << to_string(opp.type) << ": "
                      << opp.buy_source.venue << " -> " << opp.sell_source.venue
                      << " | Spread: " << opp.spread_bps.to_string() << " bps"
                      << " | Net PnL: $" << opp.net_pnl.to_string()
                      << std::endl;
        });
        std::cout << "[OK] Scanner initialized" << std::endl;

        // Initialize Cross-Chain Router
        router_ = std::make_shared<CrossChainRouter>(default_cross_chain_config());
        std::cout << "[OK] Cross-chain router initialized" << std::endl;

        // Start all systems
        lx_first_->start();
        scanner_->start();
        g_running = true;

        std::cout << std::endl;
        std::cout << std::string(60, '=') << std::endl;
        std::cout << "BOT RUNNING - Press Ctrl+C to stop" << std::endl;
        std::cout << std::string(60, '=') << std::endl;
        std::cout << std::endl;
        std::cout << "Monitoring symbols: BTC-USDC, ETH-USDC, LUX-USDC" << std::endl;
        std::cout << "Min divergence: 10 bps" << std::endl;
        std::cout << "Min profit: $5" << std::endl;
        std::cout << std::endl;

        // Start price feed simulator thread
        price_feed_thread_ = std::thread([this]() {
            simulate_price_feeds();
        });

        // Start stats reporter thread
        stats_thread_ = std::thread([this]() {
            report_stats();
        });

        return true;
    }

    void stop() {
        std::cout << "\nShutting down..." << std::endl;
        g_running = false;

        lx_first_->stop();
        scanner_->stop();

        if (price_feed_thread_.joinable()) {
            price_feed_thread_.join();
        }
        if (stats_thread_.joinable()) {
            stats_thread_.join();
        }

        print_final_stats();
    }

private:
    void on_lx_first_opportunity(const LxFirstOpportunity& opp) {
        {
            std::lock_guard<std::mutex> lock(g_stats_mutex);
            g_stats.total_opportunities++;
        }

        std::cout << std::endl;
        std::cout << std::string(50, '=') << std::endl;
        std::cout << "LX-FIRST OPPORTUNITY DETECTED" << std::endl;
        std::cout << std::string(50, '=') << std::endl;
        std::cout << "Symbol:          " << opp.symbol << std::endl;
        std::cout << "LX Price:        $" << opp.lx_price.mid.to_string() << std::endl;
        std::cout << "Stale Venue:     " << opp.stale_venue << std::endl;
        std::cout << "Stale Bid/Ask:   $" << opp.stale_price.bid.to_string()
                  << " / $" << opp.stale_price.ask.to_string() << std::endl;
        std::cout << "Staleness:       " << opp.staleness << "ms" << std::endl;
        std::cout << "Side:            " << opp.side << std::endl;
        std::cout << "Divergence:      " << opp.divergence_bps.to_string() << " bps" << std::endl;
        std::cout << "Expected Profit: $" << opp.expected_profit.to_string() << std::endl;
        std::cout << "Confidence:      " << (opp.confidence * 100) << "%" << std::endl;
        std::cout << std::string(50, '=') << std::endl;

        // Execute if confidence is high enough
        if (opp.confidence > 0.8) {
            execute_lx_first(opp);
        }
    }

    void execute_lx_first(const LxFirstOpportunity& opp) {
        std::cout << "\n[EXECUTING] " << opp.id << "..." << std::endl;

        // Determine cross-chain transport
        std::string buy_chain = router_->venue_to_chain(opp.stale_venue);
        std::string sell_chain = "lux_mainnet";
        auto transport = router_->determine_transport(buy_chain, sell_chain);
        auto latency = router_->estimate_latency(buy_chain, sell_chain);

        std::cout << "  Transport: " << to_string(transport) << std::endl;
        std::cout << "  Est. Latency: " << latency << "ms" << std::endl;

        if (opp.side == "buy") {
            std::cout << "  Buying on " << opp.stale_venue << "..." << std::endl;
            // In production: place actual order
            // auto order = cex_client->place_order(...)

            std::cout << "  Hedging on LX DEX..." << std::endl;
            // In production: place hedge order
            // auto hedge = dex_->spot()->sell(...)
        } else {
            std::cout << "  Selling on " << opp.stale_venue << "..." << std::endl;
            // In production: place actual order

            std::cout << "  Hedging on LX DEX..." << std::endl;
            // In production: place hedge order
        }

        // Simulate successful execution
        Decimal profit = opp.expected_profit * Decimal("0.8");  // Simulate slippage

        {
            std::lock_guard<std::mutex> lock(g_stats_mutex);
            g_stats.total_executions++;
            g_stats.total_pnl = g_stats.total_pnl + profit;
        }

        std::cout << "[SUCCESS] Executed " << opp.id
                  << " | Profit: $" << profit.to_string() << std::endl;
    }

    void simulate_price_feeds() {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<> price_dist(-0.001, 0.001);
        std::uniform_real_distribution<> divergence_dist(-0.002, 0.002);
        std::uniform_int_distribution<> block_dist(1000000, 2000000);

        struct BasePair {
            std::string symbol;
            Decimal base;
        };

        std::vector<BasePair> base_prices = {
            {"BTC-USDC", Decimal("50000")},
            {"ETH-USDC", Decimal("3000")},
            {"LUX-USDC", Decimal("25")}
        };

        while (g_running) {
            auto now = std::chrono::system_clock::now();
            auto timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(
                now.time_since_epoch()).count();

            for (const auto& pair : base_prices) {
                // Simulate LX DEX price (the oracle)
                double variance = price_dist(gen);
                Decimal lx_mid = pair.base * Decimal(std::to_string(1.0 + variance));

                LxPrice lx_price;
                lx_price.symbol = pair.symbol;
                lx_price.bid = lx_mid * Decimal("0.9999");
                lx_price.ask = lx_mid * Decimal("1.0001");
                lx_price.mid = lx_mid;
                lx_price.timestamp = timestamp;
                lx_price.block_num = block_dist(gen);

                lx_first_->update_lx_price(lx_price);

                // Simulate stale CEX prices
                std::vector<std::pair<std::string, int>> venues = {
                    {"binance", 50},
                    {"mexc", 100}
                };

                for (const auto& [venue, latency] : venues) {
                    double divergence = divergence_dist(gen);
                    Decimal venue_mid = pair.base * Decimal(std::to_string(1.0 + divergence));

                    VenuePrice venue_price;
                    venue_price.venue = venue;
                    venue_price.symbol = pair.symbol;
                    venue_price.bid = venue_mid * Decimal("0.9998");
                    venue_price.ask = venue_mid * Decimal("1.0002");
                    venue_price.timestamp = timestamp - latency;
                    venue_price.latency = latency;
                    venue_price.stale = false;

                    lx_first_->update_venue_price(venue_price);
                }
            }

            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
    }

    void report_stats() {
        while (g_running) {
            std::this_thread::sleep_for(std::chrono::seconds(30));

            if (!g_running) break;

            std::lock_guard<std::mutex> lock(g_stats_mutex);
            std::cout << std::endl;
            std::cout << std::string(40, '-') << std::endl;
            std::cout << "STATS" << std::endl;
            std::cout << "  Opportunities: " << g_stats.total_opportunities << std::endl;
            std::cout << "  Executions:    " << g_stats.total_executions << std::endl;
            std::cout << "  Total PnL:     $" << g_stats.total_pnl.to_string() << std::endl;
            if (g_stats.total_executions > 0) {
                Decimal avg_pnl = g_stats.total_pnl / Decimal(std::to_string(g_stats.total_executions));
                std::cout << "  Avg PnL:       $" << avg_pnl.to_string() << std::endl;
            }
            std::cout << std::string(40, '-') << std::endl;
        }
    }

    void print_final_stats() {
        std::lock_guard<std::mutex> lock(g_stats_mutex);
        std::cout << std::endl;
        std::cout << std::string(50, '=') << std::endl;
        std::cout << "FINAL STATISTICS" << std::endl;
        std::cout << std::string(50, '=') << std::endl;
        std::cout << "Total Opportunities: " << g_stats.total_opportunities << std::endl;
        std::cout << "Total Executions:    " << g_stats.total_executions << std::endl;
        std::cout << "Total PnL:           $" << g_stats.total_pnl.to_string() << std::endl;
        if (g_stats.total_executions > 0 && g_stats.total_opportunities > 0) {
            double win_rate = (static_cast<double>(g_stats.total_executions) /
                              static_cast<double>(g_stats.total_opportunities)) * 100.0;
            Decimal avg_pnl = g_stats.total_pnl / Decimal(std::to_string(g_stats.total_executions));
            std::cout << "Execution Rate:      " << win_rate << "%" << std::endl;
            std::cout << "Avg PnL per Trade:   $" << avg_pnl.to_string() << std::endl;
        }
        std::cout << std::string(50, '=') << std::endl;
    }

    static std::string get_env(const std::string& key, const std::string& default_value) {
        const char* value = std::getenv(key.c_str());
        return value ? value : default_value;
    }

    std::shared_ptr<LxDex> dex_;
    std::shared_ptr<LxFirstArbitrage> lx_first_;
    std::shared_ptr<Scanner> scanner_;
    std::shared_ptr<CrossChainRouter> router_;
    std::thread price_feed_thread_;
    std::thread stats_thread_;
};

// ============================================
// Main Entry Point
// ============================================

ArbitrageBot* g_bot = nullptr;

void signal_handler(int signal) {
    if (g_bot) {
        g_bot->stop();
    }
    std::exit(0);
}

int main() {
    // Set up signal handlers
    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    ArbitrageBot bot;
    g_bot = &bot;

    if (!bot.start()) {
        return 1;
    }

    // Keep running until signal
    while (g_running) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    return 0;
}

File: sdk/lx-trading-core/examples/arbitrage_bot.rs
//! LX-First Arbitrage Bot Example
//!
//! This bot uses the LX-First strategy where LX DEX prices are treated
//! as the "truth" (fastest venue with nanosecond updates, 200ms blocks).
//! Other venues are always stale by comparison.
//!
//! Arbitrage = exploiting stale venues before they catch up to LX prices.
//!
//! Cross-chain transport options:
//! - Warp: For Lux subnet communication (<500ms)
//! - Teleport: For EVM chain bridging (~30s)
//! - CEX API: Direct trading (instant)
//!
//! NO SMART CONTRACTS - just coordinated trades through unified SDK.

use std::sync::Arc;
use std::time::Duration;

use anyhow::Result;
use rust_decimal::Decimal;
use rust_decimal_macros::dec;
use tokio::sync::RwLock;
use tokio::time::interval;

use lx_trading::arbitrage::{
    ChainType, CrossChainConfig, CrossChainInfo, CrossChainRouter, LxFirstArbitrage, LxFirstConfig,
    LxFirstOpportunity, LxPrice, Scanner, ScannerConfig, VenuePrice,
};

// ============================================
// Arbitrage Bot
// ============================================

struct ArbitrageBot {
    lx_first: Arc<LxFirstArbitrage>,
    scanner: Arc<RwLock<Scanner>>,
    router: Arc<CrossChainRouter>,
    running: Arc<RwLock<bool>>,
    stats: Arc<RwLock<Stats>>,
}

#[derive(Default)]
struct Stats {
    total_opportunities: u64,
    total_executions: u64,
    total_pnl: Decimal,
}

impl ArbitrageBot {
    async fn new() -> Result<Self> {
        println!("{}", "=".repeat(60));
        println!("LX-FIRST ARBITRAGE BOT");
        println!("{}", "=".repeat(60));
        println!();

        // Initialize LX-First strategy
        let lx_config = LxFirstConfig {
            max_staleness_ms: 2000,
            min_divergence_bps: dec!(10),
            min_profit: dec!(5),
            max_position_size: dec!(10000),
            symbols: vec!["BTC-USDC".into(), "ETH-USDC".into(), "LUX-USDC".into()],
            venue_latencies: [
                ("binance".into(), 50),
                ("mexc".into(), 100),
                ("okx".into(), 80),
                ("uniswap".into(), 12000),
                ("pancakeswap".into(), 3000),
            ]
            .into_iter()
            .collect(),
        };
        let lx_first = Arc::new(LxFirstArbitrage::new(lx_config));
        println!("[OK] LX-First strategy initialized");

        // Initialize Scanner
        let scanner_config = ScannerConfig {
            min_spread_bps: dec!(10),
            min_profit_usd: dec!(10),
            max_price_age_ms: 5000,
            symbols: vec![
                "BTC".into(),
                "ETH".into(),
                "LUX".into(),
                "SOL".into(),
                "AVAX".into(),
            ],
            chain_ids: vec![
                "lux".into(),
                "ethereum".into(),
                "bsc".into(),
                "arbitrum".into(),
                "polygon".into(),
            ],
            scan_interval_ms: 100,
            max_concurrency: 50,
        };
        let scanner = Arc::new(RwLock::new(Scanner::new(scanner_config)));
        println!("[OK] Scanner initialized");

        // Initialize Cross-Chain Router
        let mut router_config = CrossChainConfig::default();
        router_config.warp_enabled = true;
        router_config.teleport_enabled = true;
        router_config.chains.insert(
            "lux_mainnet".into(),
            CrossChainInfo {
                chain_id: "lux_mainnet".into(),
                name: "Lux Mainnet".into(),
                chain_type: ChainType::LuxSubnet,
                block_time_ms: 400,
                finality_ms: 400,
                warp_supported: true,
                teleport_supported: true,
                venues: vec!["lx_dex".into(), "lx_amm".into()],
            },
        );
        router_config.chains.insert(
            "ethereum".into(),
            CrossChainInfo {
                chain_id: "1".into(),
                name: "Ethereum".into(),
                chain_type: ChainType::Evm,
                block_time_ms: 12000,
                finality_ms: 900000,
                warp_supported: false,
                teleport_supported: true,
                venues: vec!["uniswap".into(), "sushiswap".into()],
            },
        );
        router_config.chains.insert(
            "binance".into(),
            CrossChainInfo {
                chain_id: "binance".into(),
                name: "Binance".into(),
                chain_type: ChainType::Cex,
                block_time_ms: 0,
                finality_ms: 0,
                warp_supported: false,
                teleport_supported: false,
                venues: vec!["binance".into()],
            },
        );
        let router = Arc::new(CrossChainRouter::new(router_config));
        println!("[OK] Cross-chain router initialized");

        Ok(Self {
            lx_first,
            scanner,
            router,
            running: Arc::new(RwLock::new(false)),
            stats: Arc::new(RwLock::new(Stats::default())),
        })
    }

    async fn start(&self) -> Result<()> {
        // Start systems
        self.lx_first.start().await;
        {
            let mut scanner = self.scanner.write().await;
            scanner.start().await;
        }

        *self.running.write().await = true;

        println!();
        println!("{}", "=".repeat(60));
        println!("BOT RUNNING - Press Ctrl+C to stop");
        println!("{}", "=".repeat(60));
        println!();
        println!("Monitoring symbols: BTC-USDC, ETH-USDC, LUX-USDC");
        println!("Min divergence: 10 bps");
        println!("Min profit: $5");
        println!();

        // Spawn price feed simulator
        let lx_first = self.lx_first.clone();
        let running = self.running.clone();
        tokio::spawn(async move {
            Self::simulate_price_feeds(lx_first, running).await;
        });

        // Spawn stats reporter
        let stats = self.stats.clone();
        let running = self.running.clone();
        tokio::spawn(async move {
            Self::report_stats(stats, running).await;
        });

        Ok(())
    }

    async fn stop(&self) {
        println!("\nShutting down...");
        *self.running.write().await = false;

        self.lx_first.stop().await;
        {
            let mut scanner = self.scanner.write().await;
            scanner.stop().await;
        }

        self.print_final_stats().await;
    }

    async fn simulate_price_feeds(lx_first: Arc<LxFirstArbitrage>, running: Arc<RwLock<bool>>) {
        use rand::{rngs::StdRng, Rng, SeedableRng};

        let base_prices = [
            ("BTC-USDC", dec!(50000)),
            ("ETH-USDC", dec!(3000)),
            ("LUX-USDC", dec!(25)),
        ];

        let mut interval = interval(Duration::from_millis(100));
        let mut rng = StdRng::from_entropy();

        loop {
            interval.tick().await;

            if !*running.read().await {
                break;
            }

            for (symbol, base) in &base_prices {
                // Simulate LX DEX price (the oracle)
                let variance: f64 = rng.gen_range(-0.001..0.001);
                let lx_mid = *base * Decimal::try_from(1.0 + variance).unwrap();

                lx_first
                    .update_lx_price(LxPrice {
                        symbol: symbol.to_string(),
                        bid: lx_mid * dec!(0.9999),
                        ask: lx_mid * dec!(1.0001),
                        mid: lx_mid,
                        timestamp: chrono::Utc::now().timestamp_millis(),
                        block_num: rng.gen_range(1000000..2000000),
                    })
                    .await;

                // Simulate stale CEX prices
                for (venue, latency) in [("binance", 50i64), ("mexc", 100i64)] {
                    let divergence: f64 = rng.gen_range(-0.002..0.002);
                    let venue_mid = *base * Decimal::try_from(1.0 + divergence).unwrap();

                    lx_first
                        .update_venue_price(VenuePrice {
                            venue: venue.to_string(),
                            symbol: symbol.to_string(),
                            bid: venue_mid * dec!(0.9998),
                            ask: venue_mid * dec!(1.0002),
                            timestamp: chrono::Utc::now().timestamp_millis() - latency,
                            latency,
                            stale: false,
                        })
                        .await;
                }
            }
        }
    }

    async fn handle_opportunity(
        opp: &LxFirstOpportunity,
        router: &CrossChainRouter,
        stats: &Arc<RwLock<Stats>>,
    ) {
        println!();
        println!("{}", "=".repeat(50));
        println!("LX-FIRST OPPORTUNITY DETECTED");
        println!("{}", "=".repeat(50));
        println!("Symbol:          {}", opp.symbol);
        println!("LX Price:        ${}", opp.lx_price.mid);
        println!("Stale Venue:     {}", opp.stale_venue);
        println!(
            "Stale Bid/Ask:   ${} / ${}",
            opp.stale_price.bid, opp.stale_price.ask
        );
        println!("Staleness:       {}ms", opp.staleness);
        println!("Side:            {}", opp.side.to_uppercase());
        println!("Divergence:      {} bps", opp.divergence_bps);
        println!("Expected Profit: ${}", opp.expected_profit);
        println!("Confidence:      {:.1}%", opp.confidence * 100.0);
        println!("{}", "=".repeat(50));

        if opp.confidence > 0.8 {
            println!("\n[EXECUTING] {}...", opp.id);

            // Determine cross-chain transport
            let buy_chain = router.venue_to_chain(&opp.stale_venue);
            let sell_chain = "lux_mainnet";
            let transport = router.determine_transport(&buy_chain, sell_chain);
            let latency = router.estimate_latency(&buy_chain, sell_chain);

            println!("  Transport: {transport:?}");
            println!("  Est. Latency: {latency}ms");

            if opp.side == "buy" {
                println!("  Buying on {}...", opp.stale_venue);
                println!("  Hedging on LX DEX...");
            } else {
                println!("  Selling on {}...", opp.stale_venue);
                println!("  Hedging on LX DEX...");
            }

            // Simulate successful execution
            let profit = opp.expected_profit * dec!(0.8);

            {
                let mut s = stats.write().await;
                s.total_executions += 1;
                s.total_pnl += profit;
            }

            println!("[SUCCESS] Executed {} | Profit: ${:.2}", opp.id, profit);
        }

        {
            let mut s = stats.write().await;
            s.total_opportunities += 1;
        }
    }

    async fn report_stats(stats: Arc<RwLock<Stats>>, running: Arc<RwLock<bool>>) {
        let mut interval = interval(Duration::from_secs(30));

        loop {
            interval.tick().await;

            if !*running.read().await {
                break;
            }

            let s = stats.read().await;
            println!();
            println!("{}", "-".repeat(40));
            println!("STATS");
            println!("  Opportunities: {}", s.total_opportunities);
            println!("  Executions:    {}", s.total_executions);
            println!("  Total PnL:     ${:.2}", s.total_pnl);
            if s.total_executions > 0 {
                let avg_pnl = s.total_pnl / Decimal::from(s.total_executions);
                println!("  Avg PnL:       ${avg_pnl:.2}");
            }
            println!("{}", "-".repeat(40));
        }
    }

    async fn print_final_stats(&self) {
        let s = self.stats.read().await;
        println!();
        println!("{}", "=".repeat(50));
        println!("FINAL STATISTICS");
        println!("{}", "=".repeat(50));
        println!("Total Opportunities: {}", s.total_opportunities);
        println!("Total Executions:    {}", s.total_executions);
        println!("Total PnL:           ${:.2}", s.total_pnl);
        if s.total_executions > 0 && s.total_opportunities > 0 {
            let win_rate = (s.total_executions as f64 / s.total_opportunities as f64) * 100.0;
            let avg_pnl = s.total_pnl / Decimal::from(s.total_executions);
            println!("Execution Rate:      {win_rate:.1}%");
            println!("Avg PnL per Trade:   ${avg_pnl:.2}");
        }
        println!("{}", "=".repeat(50));
    }
}

// ============================================
// Main Entry Point
// ============================================

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize logging
    tracing_subscriber::fmt::init();

    let bot = ArbitrageBot::new().await?;
    bot.start().await?;

    // Wait for Ctrl+C
    tokio::signal::ctrl_c().await?;
    bot.stop().await;

    Ok(())
}
